library/core/src/gc.rs

```diff
@@ -25,6 +37,27 @@ pub struct Trace {
     pub size: u64,
 }
 
+/// A wrapper which prevents `T` from being finalized when used in a `Gc`.
+///
+/// This has the same effect as implementing `NoFinalize` trait on `T`, however,
+/// due to the orphan rule this is not always possible. `NonFinalizable` acts as
+/// a convenience wrapper.
+#[derive(Debug)]
+pub struct NonFinalizable<T: ?Sized>(T);
+
+impl<T> NonFinalizable<T> {
+    /// Wrap a value to prevent finalization in `Gc`.
+    ///
+    /// # Safety
+    ///
+    /// Unsafe because this should be used with care. Preventing drop from
+    /// running can lead to surprising behaviour. In particular, this will also
+    /// prevent the finalization of all component types of T.
+    pub unsafe fn new(value: T) -> NonFinalizable<T> {
```

Comment:
Just to be awkward, since this is `ManuallyDrop` in disguise, I think it's OK for this to not be marked as `unsafe` (`ManuallyDrop::new` is not `unsafe`). I don't feel strongly about that though!